### PR TITLE
Removing lenses

### PR DESCRIPTION
**Pull request by Heimdell**
- [x] Remove lenses
- [x] Remove `mlNextKey`/`mlPrevKey`
- [x] Make `mlHash` lazy and `h` instead of `Maybe h`, perform lazy rehash on each change
- [x] Fix tests.
- [ ] Stylish-haskell.

### CODE REVIEW

File: src/Data/Tree/AVL.hs
Comment (kevroletin) [inline src/Data/Tree/AVL.hs]: What usecases are assumed for beforeSerialise, afterDeserialise and mapTilt? There where discussion that AVL+ API try to hide that AVL+ is a Tree so I assume that Tilt is exposed for some important use case

**Review by kevroletin — COMMENTED** [review]
I am suggesting adding comments to commits.

**Comment by kevroletin** [comment]
There was also a conversation about new requirements to a user-defined hash function. Can we document them?

File: src/Data/Tree/AVL.hs
Comment (Heimdell) [reply src/Data/Tree/AVL.hs]: `beforeSerialise` turns `Tilt`s into `Int`s, so you can the serialise the result and don't write any instances for `Tilt`.

`afterDeserialise` turns `Int`s into `Tilt`.

File: config.yaml
Comment (flyingleafe) [inline config.yaml]: Seems like this file has been added by mistake.

File: src/Data/Tree/AVL/Internal.hs
Comment (flyingleafe) [inline src/Data/Tree/AVL/Internal.hs]: 1) I really don't think that writing `Serialise` instances for `Tilt` is such an ordeal that we really need all those helpers. Especially given that it seems that all the `*Template` type aliases are needed only for this.
2) Even if we have them, `Tilt` should be translated into `Word8` instead of `Int` (why on earth we would need 32 bits for 5 values)

File: src/Data/Tree/AVL/Zipper.hs
Comment (flyingleafe) [inline src/Data/Tree/AVL/Zipper.hs:75]: Lol, memno

File: src/Data/Tree/AVL/Zipper.hs
Comment (flyingleafe) [inline src/Data/Tree/AVL/Zipper.hs:166]: Again, why not just delete that

**Review by flyingleafe — CHANGES_REQUESTED** [review]
Reviewed partially. Will review again after #13 is merged and changes introduced there are isolated.

File: src/Data/Tree/AVL/Zipper.hs
Comment (Heimdell) [reply src/Data/Tree/AVL/Zipper.hs:75]: It was a long-standing bug.

File: src/Data/Tree/AVL/Zipper.hs
Comment (Heimdell) [reply src/Data/Tree/AVL/Zipper.hs:166]: I won't delete these two functions, because each time I change something and something other breaks, I have to us'em to find what went wrong.

Comment (Heimdell) [reply src/Data/Tree/AVL/Zipper.hs:166]: I'm about `dump` and `say`.

File: test/Deletion.hs
Comment (Martoon-00) [inline test/Deletion.hs]: `dpr`? :smile_cat:

File: config.yaml
Comment (Heimdell) [reply config.yaml]: Yep.

File: src/Data/Tree/AVL/Internal.hs
Comment (Heimdell) [reply src/Data/Tree/AVL/Internal.hs]: You're right in (2).

I think, `Tilt` is an... look, I can literally replace `Tilt` with `Word8`. This fixes both problems: 1) removes `Tilt` as API part; 2) removes `*Template`s.

File: test/Deletion.hs
Comment (Heimdell) [reply test/Deletion.hs]: This is so I was able to rerun the test with `--test-arguments --match=dpr`

File: src/Data/Tree/AVL/Zipper.hs
Comment (flyingleafe) [reply src/Data/Tree/AVL/Zipper.hs:166]: @Heimdell well then just have some special place for them (like `Data.Tree.AVL.Debug`) and don't comment them, just leave them alone.

File: src/Data/Tree/AVL/Insertion.hs
Comment (flyingleafe) [inline src/Data/Tree/AVL/Insertion.hs]: oof, why newline and additional 2-space tab after `->`? one line was better by all means.

File: src/Data/Tree/AVL/Store/Pure.hs
Comment (flyingleafe) [inline src/Data/Tree/AVL/Store/Pure.hs:44]: I know very well that you're a fan of vertical alignment, but let's just consider another opinion: http://www.zeyalabs.ch/posts/2013/vertical-alignment-sucks/

File: src/Data/Tree/AVL/Zipper.hs
Comment (flyingleafe) [inline src/Data/Tree/AVL/Zipper.hs]: What's the purpose of this rename?

File: test/Deletion.hs
Comment (flyingleafe) [inline test/Deletion.hs]: Please, don't forget to remove your pattern shortcuts from the descriptions. Also, you can just use `focus` temporarily if you need to debug one particular test.

**Review by flyingleafe — APPROVED** [review]
LGTM, but it would still be nice if you address my minor, mostly stylistic comments.

File: src/Data/Tree/AVL/Store/Pure.hs
Comment (Heimdell) [reply src/Data/Tree/AVL/Store/Pure.hs:44]: Readability of code is of higher concern for me, so that one I'll not undo.

File: src/Data/Tree/AVL/Insertion.hs
Comment (Heimdell) [reply src/Data/Tree/AVL/Insertion.hs]: Fix'd

File: src/Data/Tree/AVL/Zipper.hs
Comment (Heimdell) [reply src/Data/Tree/AVL/Zipper.hs]: Fix'd

File: test/Deletion.hs
Comment (Heimdell) [reply test/Deletion.hs]: Fix'd

**Pull request merged**
